internal/db/p2p/sync_branchable_col.go

```
@@ -0,0 +1,335 @@
+
+	"github.com/sourcenetwork/corelog"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/event"
+	coreblock "github.com/sourcenetwork/defradb/internal/core/block"
+	"github.com/sourcenetwork/defradb/internal/datastore"
+	dbid "github.com/sourcenetwork/defradb/internal/db/id"
+	"github.com/sourcenetwork/defradb/internal/keys"
+)
+
+// syncBranchableCollectionTopic is the fixed topic for branchable collection sync operations.
+const syncBranchableCollectionTopic = "sync-branchable"
+
+// syncBranchableCollectionRequest represents a request to synchronize a branchable collection.
+type syncBranchableCollectionRequest struct {
+	CollectionID string `json:"collectionID"`
+}
+
+// syncBranchableCollectionReply represents the response to a collection sync request.
+type syncBranchableCollectionReply struct {
+	CollectionID string   `json:"collectionID"`
+	Heads        [][]byte `json:"heads"`
+	Sender       string   `json:"sender"`
+}
+
+// SyncBranchableCollection initiates a request for the latest version of the branchable
+// collection's DAG from the network.
+//
+// This function call will block until there is a response for the collection.
+// It is the responsibility of the caller to set an appropriate timeout on the context.
+func (p *P2P) SyncBranchableCollection(ctx context.Context, collectionID string) error {
+	log.InfoContext(ctx, "Starting branchable collection sync",
+		corelog.String("PeerID", p.host.ID()))
+
+	cols, err := p.db.GetCollections(
+		ctx,
+		client.CollectionFetchOptions{
+			CollectionID: immutable.Some(collectionID),
+		},
+	)
+	if err != nil {
+		return err
+	}
+	if len(cols) == 0 {
+		return client.NewErrCollectionNotFoundForCollectionVersion(collectionID)
+	}
+
+	col := cols[0].Version()
+	if !col.IsBranchable {
+		return errors.New("collection is not branchable", errors.NewKV("CollectionID", collectionID))
```

todo: Please return an error from `errors.go` instead of declaring inline like this.

---

done